display_list/display_list_unittests.cc

```diff
@@ -2499,6 +2504,296 @@ TEST(DisplayList, RTreeOfSaveLayerFilterScene) {
+    builder2.drawRect({10, 10, 20, 20});
+    builder2.save();
+    builder2.translate(1.0, 1.0);
+    { builder2.drawRect({50, 50, 60, 60}); }
+    builder2.restore();
+    ASSERT_TRUE(DisplayListsEQ_Verbose(builder.Build(), builder2.Build()));
+  }
+}
+
+TEST(DisplayList, CollapseMultipleNestedSaveRestore) {
+  DisplayListBuilder builder1;
+  builder1.save();
+  builder1.save();
+  builder1.save();
+  builder1.translate(10, 10);
+  builder1.scale(2, 2);
+  builder1.clipRect({10, 10, 20, 20}, SkClipOp::kIntersect, false);
+  builder1.drawRect({0, 0, 100, 100});
+  builder1.restore();
+  builder1.restore();
+  builder1.restore();
+  auto display_list1 = builder1.Build();
+
+  DisplayListBuilder builder2;
+  builder2.save();
+  builder2.translate(10, 10);
+  builder2.scale(2, 2);
+  builder2.clipRect({10, 10, 20, 20}, SkClipOp::kIntersect, false);
+  builder2.drawRect({0, 0, 100, 100});
+  builder2.restore();
+  auto display_list2 = builder2.Build();
+
+  ASSERT_TRUE(DisplayListsEQ_Verbose(display_list1, display_list2));
+}
+
+TEST(DisplayList, RemoveUnnecessarySaveRestorePairsInSetPaint) {
+  SkRect build_bounds = SkRect::MakeLTRB(-100, -100, 200, 200);
+  SkRect rect = SkRect::MakeLTRB(30, 30, 70, 70);
+  // clang-format off
+  const float alpha_matrix[] = {
+    0, 0, 0, 0, 0,
+    0, 1, 0, 0, 0,
+    0, 0, 1, 0, 0,
+    0, 0, 0, 0, 1,
+  };
+  // clang-format on
+  DlMatrixColorFilter alpha_color_filter(alpha_matrix);
+  // Making sure hiding a problematic ColorFilter as an ImageFilter
+  // will generate the same behavior as setting it as a ColorFilter
+
+  DlColorFilterImageFilter color_filter_image_filter(alpha_color_filter);
+  {
+    DisplayListBuilder builder(build_bounds);
+    builder.save();
+    DlPaint paint;
+    paint.setImageFilter(&color_filter_image_filter);
+    builder.drawRect(rect, paint);
+    builder.restore();
+    sk_sp<DisplayList> display_list1 = builder.Build();
+
+    DisplayListBuilder builder2(build_bounds);
+    DlPaint paint2;
+    paint2.setImageFilter(&color_filter_image_filter);
+    builder2.drawRect(rect, paint2);
+    sk_sp<DisplayList> display_list2 = builder2.Build();
+    ASSERT_TRUE(DisplayListsEQ_Verbose(display_list1, display_list2));
+  }
+
+  {
+    DisplayListBuilder builder(build_bounds);
+    builder.save();
+    builder.saveLayer(&build_bounds, true);
+    DlPaint paint;
+    paint.setImageFilter(&color_filter_image_filter);
+    builder.drawRect(rect, paint);
+    builder.restore();
+    builder.restore();
+    sk_sp<DisplayList> display_list1 = builder.Build();
+
+    DisplayListBuilder builder2(build_bounds);
+    builder2.saveLayer(&build_bounds, true);
+    DlPaint paint2;
+    paint2.setImageFilter(&color_filter_image_filter);
+    builder2.drawRect(rect, paint2);
+    builder2.restore();
+    sk_sp<DisplayList> display_list2 = builder2.Build();
+    ASSERT_TRUE(DisplayListsEQ_Verbose(display_list1, display_list2));
+  }
+}
+
+TEST(DisplayList, CollapseNestedSaveAndSaveLayerRestore) {
```

Comment:
There seems to be some confusion over the way the following code blocks in this test are organized.

First, this test is named "CollapseNestedSaveAndSaveLayerRestore" which matches the scenario tested in the first code block - it is verifying that we collapse a save and a saveLayer together.

But, that first code block is somehow labeled "Translate" which seems like an arbitrary and misleading description of what it is doing - possibly related to the rest of the test-code-blocks here.

The rest of the code blocks appear to be verifying that each of the state-changing methods trigger a deferred save and are labeled with comments such as Transform, Skew, Scale, etc.

Let's break this all up and fix the comments - actually deleting the comments would be my recommendation - instead having the various tests for the methods that trigger a deferred save each be in their own top level "TEST(...)" with the appropriate name such as:

```
TEST(DisplayList, CollapseNestedSaveAndSaveLayerRestore) {
  // the first code block here
}
TEST(DisplayList, TranslateTriggersDeferredSave) {...}
TEST(DisplayList, ScaleTriggersDeferredSave) {...}
TEST(DisplayList, SkewTriggersDeferredSave) {...}
TEST(DisplayList, Transform2DTriggersDeferredSave) {...}
TEST(DisplayList, TransformPerspectiveTriggersDeferredSave) {...}
TEST(DisplayList, ResetTransformTriggersDeferredSave) {...}
TEST(DisplayList, ClipRectTriggersDeferredSave) {...}
TEST(DisplayList, ClipRRectTriggersDeferredSave) {...}
TEST(DisplayList, ClipPathTriggersDeferredSave) {...}
```

and perhaps also have:

```
TEST(DisplayList, NopTranslateDoesNotTriggerDeferredSave) {...}
// ... etc.
```

where we verify that translate(0, 0), and other NOP transform (and clip?) calls that check their arguments, don't bother emitted a deferred save/restore.